internal/abi/fastly/hostcalls_guest.go

```diff
@@ -636,13 +638,19 @@ func fastlyHTTPReqDownstreamTLSClientHello(
 // DownstreamTLSClientHello returns the ClientHello message sent by the client
 // in the singleton downstream request, if any.
 func DownstreamTLSClientHello() ([]byte, error) {
-	buf := prim.NewWriteBuffer(defaultBufferLen)
-
-	if err := fastlyHTTPReqDownstreamTLSClientHello(
+	n := DefaultLargeBufLen
+alloc:
+	buf := prim.NewWriteBuffer(n) // Longest (~132,000); typically < 2^14; RFC https://datatracker.ietf.org/doc/html/rfc8446#section-4.1.2
+	status := fastlyHTTPReqDownstreamTLSClientHello(
 		prim.ToPointer(buf.Char8Pointer()),
 		buf.Cap(),
 		prim.ToPointer(buf.NPointer()),
-	).toError(); err != nil {
+	)
+	if status == FastlyStatusBufLen && buf.NValue() > 0 {
+		n = int(buf.NValue())
+		goto alloc // goto saves having to allocate a function closure and avoids having to duplicate the hostcall
+	}
+	if err := status.toError(); err != nil {
 		return nil, err
 	}
 
```

Comment:
I'm fine with the `goto`, but maybe this could be an infinite loop instead? Then there's no need to explain why you're using it.

```go
for {
     buf := ...
     status := ...
     if status == FastlyStatusBufLen && buf.NValue() > 0 {
             n = int(buf.NValue())
             continue
     }
     if err := ... {
             return nil, err
     }

     return buf.AsBytes(), nil
}
```

Reply:
Here's an attempt for `Dictionary#GetBytes()` (it only partially support adaptive buffers with a known maximum size):
``` go
// Get the value for key, as a byte slice, if it exists.
func (d *Dictionary) GetBytes(key string) ([]byte, error) {
	keyBuffer := prim.NewReadBufferFromString(key).Wstring()
	n := DefaultMediumBufLen // Longest (8192) = Config Store limit; typical values likely less than 1024
	for {
		buf := prim.NewWriteBuffer(n)
		status := fastlyDictionaryGet(
			d.h,
			keyBuffer.Data, keyBuffer.Len,
			prim.ToPointer(buf.Char8Pointer()), buf.Cap(),
			prim.ToPointer(buf.NPointer()),
		)
		if status == FastlyStatusBufLen && n < dictionaryValueMaxLen {
			// The Dictionary API cannot return the needed size with this error.
			// Instead of perfectly adapting, we allocate the maximum length a value can have.
			n = dictionaryValueMaxLen
			continue
		}
		if err := status.toError(); err != nil {
			return nil, err
		}
		return buf.AsBytes(), nil
	}
}
```

Reply:
Hard to say if this produces more allocations with the `:=` in the loop than the goto version. I'm not even sure how to test that.

Reply:
It shouldn't be different! `for {}` is just a `goto` and a label in a trenchcoat.

Reply:
The original writebuf allocation will be lost if an reallocation was needed, so maybe we could also have a buffer pool to reuse allocations made with adaptive buffer sizes that had to be reallocated?

Reply:
Yep, pool will come in a following PR

Reply:
Thanks for the comments. Improved the end result!